flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -348,20 +353,41 @@ private void handleStartTaskExecutorServicesException(Exception e) throws Except
 		resourceManagerHeartbeatManager.stop();
 
 		try {
-			stopTaskExecutorServices();
+			stopServicesBeforeTasksCompletion();
 		} catch (Exception e) {
 			throwable = ExceptionUtils.firstOrSuppressed(e, throwable);
 		}
 
 		if (throwable != null) {
 			return FutureUtils.completedExceptionally(new FlinkException("Error while shutting the TaskExecutor down.", throwable));
 		} else {
-			log.info("Stopped TaskExecutor {}.", getAddress());
-			return CompletableFuture.completedFuture(null);
+			return FutureUtils.waitForAll(taskCompletionFutures).thenRun(() -> {
+				try {
+					stopServicesAfterTasksCompletion();
+				} catch (Exception e) {
+					throw new FlinkRuntimeException("Error while shutting the TaskExecutor down.", e);
+				}
+				log.info("Stopped TaskExecutor {}.", getAddress());
+			});
 		}
 	}
 
-	private void stopTaskExecutorServices() throws Exception {
+	private List<CompletableFuture<Void>> failTasks(JobID jobID, Throwable cause) {
+		List<CompletableFuture<Void>> taskCompletionFutures = new ArrayList<>();
+		Iterator<Task> tasks = taskSlotTable.getTasks(jobID);
+		while (tasks.hasNext()) {
+			try {
+				Task task = tasks.next();
+				task.failExternally(cause);
+				taskCompletionFutures.add(task.getTaskCompletionFuture());
+			} catch (Throwable t) {
+				taskCompletionFutures.add(FutureUtils.completedExceptionally(t));
+			}
+        }
+        return taskCompletionFutures;
+	}
+
+	private void stopServicesBeforeTasksCompletion() throws Exception {
```

Comment:
Why did you separate the service shutdown into two stages? Can't we simply stop all services after the `Tasks` have been terminated?